src/aids/forms.py

```diff
@@ -687,6 +689,42 @@ def perimeter_filter(self, qs, search_perimeter):
         qs = qs.filter(perimeter__in=perimeter_qs)
         return qs
 
+    def typology_filter(self, qs, search_perimeter):
+        """
+        We should never have both the generic aid and it's local version
+        together on search results.
+        Which one should be removed from the result ? It depends...
+        We consider the scale perimeter associated to the local aid.
+        - When searching a on wider area than the local aid's perimeter,
+          then we display the generic version.
+        - When searching on a smaller area than the local aid's perimeter,
+          then we display the local version.
+        """
+        local_aids = qs.filter(aid_typology=Aid.LOCAL_TYPOLOGY)
+        aids_to_exclude = []
+        if not search_perimeter:
+            # If the user does not specify a search perimeter, then we go wide.
+            search_smaller = False
+            search_wider = True
+        for aid in local_aids:
+            if search_perimeter:
+                search_smaller = search_perimeter.scale <= aid.perimeter.scale
+                search_wider = search_perimeter.scale > aid.perimeter.scale
+            generic_aid_is_present = aid.generic_aid and aid.generic_aid in qs
```

Comment:
C'est encore un peu flou dans mon esprit :

 - `aid.generic_aid` : si une aide est de typologie « locale », est-ce qu'il n'est pas obligatoire de lui associer une aide générique ?
 - `aid.generic_aid in qs` : si une aide locale est renvoyée par la recherche, est-ce qu'on ne part pas du principe que la recherche générique est également présente ?

Reply:
`si une aide est de typologie « locale », est-ce qu'il n'est pas obligatoire de lui associer une aide générique ?`
Oui, tu proposes de rajouter une contrainte ? J'avais pensé à ces deux contraintes - à voir s'il faut le faire dans une autre PR:
    - Forcer les aides locales à avoir une FK aide générique.
    - S'assurer que seule les aides locales peuvent renseigner la FK aide générique.

Reply:
`si une aide locale est renvoyée par la recherche, est-ce qu'on ne part pas du principe que la recherche générique est également présente`
Ça parait logique, oui, mais je ne sais pas encore trop comment on va utiliser ces aides locales. Et si par exemple on fait une recherche textuelle qui fait qu'on ne sélectionne que l'aide locale et pas l'aide générique. Je vais demander clarification pour voir si on peut s'économiser cette vérification `aid.generic_aid in qs`.

Reply:
>  Oui, tu proposes de rajouter une contrainte ? J'avais pensé à ces deux contraintes - à voir s'il faut le faire dans une autre PR:

Autre question : est-ce qu'il est vraiment obligatoire d'indiquer à la main la typologie de l'aide ? Si on ne gardait que la clé, est-ce que ça ne suffirait pas ?

Si ton aide pointe vers une autre aide : elle est locale.
Si ton aide est pointée par une autre : elle est globale.

Reply:
Oui je pense que c'est possible de faire comme tu le décris, je n'y avais pas pensé. Ça paraissait plus facile d'avoir un champs explicite. Je sais pas ce qui est le mieux.

Reply:
pas mal en effet d'avoir juste un champ qui stock l'info 👌

Reply:
et ca permettra peut-etre de faire tout en filter querying sans avoir à faire de for loop

Reply:
Je vais proposer ces changements dans une nouvelle PR.

Reply:
J'ai fait une proposition de simplification ici : https://github.com/MTES-MCT/aides-territoires/pull/441